### PR TITLE
Add geo fields to `add_host_metadata` processor.

### PR DESCRIPTION
**Pull request by andrewvc**
** EDIT ** I've left the original issue below the break, but after discussion we added geo fields to the `add_host_metadata` processor instead of a new one. *Original is below*

-----

This carries over from the discussion from https://github.com/elastic/beats/pull/8620 .

This adds a new processor that lets users easily add geo fields associated with the host that created the event.  You would use it like so:

```
processors:
  - add_host_geo:
      name: MN HQ
      location: "44.977753, -93.265015"
```

It's debate-able whether ECS should actually let you put these under perhaps `agent.geo`. That's something we should discuss here.

One other question here, should we just fold this functionality under `add_host_metadata`? I believe that probably makes more sense. We agreed in elastic/beats#8620 to make this a separate processor, but with the data nested under host, that makes less sense IMHO.

### CODE REVIEW

**Comment by andrew-moldovan** [comment]
I believe I was added as a reviewer by mistake :) Just pointing it out in case you meant to tag somebody else 😄

File: libbeat/processors/add_host_geo/add_host_geo.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+type addGeo common.MapStr
+
+func (ag addGeo) Run(event *beat.Event) (*beat.Event, error) {
+	_, err := event.PutValue("host.geo", ag.asMap())
+	return event, err
+}
+
+func (ag addGeo) asMap() common.MapStr {
+	return common.MapStr(ag).Clone()
```
Comment (urso) [inline libbeat/processors/add_host_geo/add_host_geo.go]: nit: +1 on cloning, but from name only I would asMap expect to be a cast operation only. Caught me a little offguard when reading the argument for PutValue.

File: libbeat/processors/add_host_geo/add_host_geo.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+type Config struct {
+	Name           string `config:"name"`
+	Location       string `config:"location"`
+	ContinentName  string `config:"continent_name"`
+	CountryISOCode string `config:"country_iso_code"`
+	RegionName     string `config:"region_name"`
+	CityName       string `config:"city_name"`
+}
```
Comment (urso) [inline libbeat/processors/add_host_geo/add_host_geo.go]: I'd prefer to have validation for at least the fields that are stored as geo point in Elasticsearch here.

Comment (andrewvc) [reply libbeat/processors/add_host_geo/add_host_geo.go]: Ah, so we'd check the format to be a lat lon pair? I think that makes sense. Anything beyond that?

File: libbeat/processors/add_host_geo/add_host_geo.go (hunk 1 above)
Comment (andrewvc) [reply libbeat/processors/add_host_geo/add_host_geo.go]: Good point, I'll rename to `mapClone()`.

File: libbeat/_meta/fields.ecs.yml [hunk 3]
```diff
@@ -259,6 +259,11 @@
             description: >
               Region ISO code.
 
+          - name: name
```
Comment (ruflin) [inline libbeat/_meta/fields.ecs.yml]: Why is there such a large diff? Did you replace the ECS yml? Please don't do this in this PR.

File: libbeat/processors/add_host_geo/add_host_geo.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	processors.RegisterPlugin("add_host_geo", newAddGeo)
+}
+
+func newAddGeo(c *common.Config) (processors.Processor, error) {
+	config := &Config{}
+	c.Unpack(config)
+
+	ag := addGeo{
+		"name":             config.Name,
+		"location":         config.Location,
```
Comment (ruflin) [inline libbeat/processors/add_host_geo/add_host_geo.go]: Should we / can we do any validation for the location field to make sure we don't have ingestion errors?

File: libbeat/processors/add_host_geo/add_host_geo.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	for k, v := range ag {
+		if v == "" {
+			delete(ag, k)
+		}
+	}
+
+	return ag, nil
+}
+
+type Config struct {
```
Comment (ruflin) [inline libbeat/processors/add_host_geo/add_host_geo.go]: Nit but I'm somehow used to have configs on top

File: libbeat/processors/add_host_geo/add_host_geo.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	ContinentName  string `config:"continent_name"`
+	CountryISOCode string `config:"country_iso_code"`
+	RegionName     string `config:"region_name"`
+	CityName       string `config:"city_name"`
+}
+
+type addGeo common.MapStr
+
+func (ag addGeo) Run(event *beat.Event) (*beat.Event, error) {
+	_, err := event.PutValue("host.geo", ag.asMap())
```
Comment (ruflin) [inline libbeat/processors/add_host_geo/add_host_geo.go]: Does this really belong into the host? We are interested where the agent is running so it would be `agent.geo`?

File: libbeat/processors/add_host_geo/add_host_geo.go (hunk 2 above)
Comment (urso) [reply libbeat/processors/add_host_geo/add_host_geo.go]: yes, at least lat-lon pair.

The others are kind free form text and validating these might require a location database, right? We even store the others as keyword, so not that important. Validating them might be too much effort here (would be nice if we can do some simple validation - regex like -, but no hard requirement for me).

Most important is that indexing won't fail due to misconfiguration of the processor + maps in kibana dashboards will work correctly, cause lat lon pairs are correct and make sense.

File: libbeat/processors/add_host_geo/add_host_geo.go (hunk 6 above)
Comment (andrewvc) [reply libbeat/processors/add_host_geo/add_host_geo.go]: It's an interesting question. I agree that agent is slightly better, but I wonder if that is outweighed by the simplicity of not having yet another geo field. A location is tied to a host as much as it is tied to an agent.

There's an argument that's not true for the `geo.name` field we use, which is free-form, but perhaps we don't need to be so dogmatic there.

Comment (ruflin) [reply libbeat/processors/add_host_geo/add_host_geo.go]: One point that would speak for the host is that in the case of apm-server, apm-server is not the agent but the observer. There it would then be `observer.geo.*`. Having it under `host` would then not change it. I'm good with going with `host.*` for now and if someone is not happy with that, he can use the `rename` processor.

Comment (MikePaquette) [reply libbeat/processors/add_host_geo/add_host_geo.go]: +1 for **not** having `agent.geo.*`
As @ruflin says, in ECS, the `agent` is defined as running on a `host` or an `observer`.

Both `host.geo.*` and `observer.geo.*` are defined by ECS, so we are covered in either case.

**Comment by andrewvc** [comment]
After a discussion with @ruflin we decided to move this to the `add_host_metadata` processor

**Comment by andrewvc** [comment]
OK, redid the whole thing under the `add_host_metadata` processor.

The only thing not yet done is the changes to fields.yml. Since this is now in ECS, is there a process for updated the ECS fields, or should I just manually add the ECS name field?

File: libbeat/processors/add_host_geo/add_host_geo.go (hunk 2 above)
Comment (andrewvc) [reply libbeat/processors/add_host_geo/add_host_geo.go]: Validation added :)

File: libbeat/processors/add_host_metadata/add_host_metadata.go [hunk 7]
```diff
@@ -62,6 +64,32 @@ func newHostMetadataProcessor(cfg *common.Config) (processors.Processor, error)
 		data:   common.NewMapStrPointer(nil),
 	}
 	p.loadData()
+
+	if config.Geo != nil {
+		if len(config.Geo.Location) > 0 {
+			if m, _ := regexp.MatchString("^\\-?\\d+(\\.\\d+)?\\s*\\,\\s*\\-?\\d+(\\.\\d+)?$", config.Geo.Location); !m {
```
Comment (ruflin) [inline libbeat/processors/add_host_metadata/add_host_metadata.go]: I'm undecided if I like this one or not. It's good that we check the user input but it looks like really complex regexp where I must confess I could not easily review to make sure it's correct.

Comment (andrewvc) [reply libbeat/processors/add_host_metadata/add_host_metadata.go]: I can break the regexp up into component strings to make ti more readable.

Comment (andrewvc) [reply libbeat/processors/add_host_metadata/add_host_metadata.go]: FYI, the regexp isn't 100% accurate (it doesn't check the numeric bounds), but it's a lot better than nothing

Comment (andrewvc) [reply libbeat/processors/add_host_metadata/add_host_metadata.go]: In b4feb7e I've made the regex more clear

**Comment by ruflin** [comment]
@andrewvc Feel free to edit the ecs or the libbeat common file for the field, both works.

**Comment by andrewvc** [comment]
Looks like rebasing of master gives us the geo fields thanks to @andrewkroh :) in 66a952

File: libbeat/processors/add_host_metadata/add_host_metadata.go [hunk 8]
```diff
@@ -62,6 +64,44 @@ func newHostMetadataProcessor(cfg *common.Config) (processors.Processor, error)
 		data:   common.NewMapStrPointer(nil),
 	}
 	p.loadData()
+
+	if config.Geo != nil {
+		if len(config.Geo.Location) > 0 {
+			// Regexp matching a number with an optional decimal component
+			// Valid numbers: '123', '123.23', etc.
+			latOrLon := `\-?\d+(\.\d+)?`
```
Comment (ruflin) [inline libbeat/processors/add_host_metadata/add_host_metadata.go]: Nice. Thanks for all the details here.

File: libbeat/processors/add_host_metadata/add_host_metadata_test.go [hunk 9]
```diff
@@ -112,3 +115,69 @@ func TestConfigNetInfoEnabled(t *testing.T) {
...
+			_, err = newHostMetadataProcessor(conf)
+
+			if location.valid {
+				require.NoError(t, err)
+			} else {
+				require.Error(t, err)
+			}
+		})
+	}
+}
```
Comment (webmat) [inline libbeat/processors/add_host_metadata/add_host_metadata_test.go]: Nice tests.

Perhaps one more to test "Delete any empty values"? Make sure your empty values include things like empty spaces too `{"city_name": "   "}`

Comment (andrewvc) [reply libbeat/processors/add_host_metadata/add_host_metadata_test.go]: @webmat I've added handling + tests for blank strings. Mind taking a look?

Comment (webmat) [reply libbeat/processors/add_host_metadata/add_host_metadata_test.go]: Yeah good stuff.

Seems like I noticed the `"din expect"` typo too late for this PR ;-)
